How to call `CreateFromMemory`?

The [`ktxTexture2_CreateFromMemory`](https://github.com/KhronosGroup/KTX-Software/blob/7a45c4e6159c0815e4663a529a1e718136a16e39/include/ktx.h#L1055) function has the following signature:
```c
ktxTexture2_CreateFromMemory(const ktx_uint8_t* bytes, ktx_size_t size,
                             ktxTextureCreateFlags createFlags,
                             ktxTexture2** newTex)

```
It receives `ktxTextureCreateFlags`. These are typedef'ed to `ktx_uint32_t`, and there is an enum
```c
enum ktxTextureCreateFlagBits {
    KTX_TEXTURE_CREATE_NO_FLAGS = 0x00,
    KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT = 0x01,
                                   /*!< Load the images from the KTX source. */
    KTX_TEXTURE_CREATE_RAW_KVDATA_BIT = 0x02,
                                   /*!< Load the raw key-value data instead of
                                        creating a @c ktxHashList from it. */
    KTX_TEXTURE_CREATE_SKIP_KVDATA_BIT = 0x04,
                                   /*!< Skip any key-value data. This overrides
                                        the RAW_KVDATA_BIT. */
    KTX_TEXTURE_CREATE_CHECK_GLTF_BASISU_BIT = 0x08
                                   /*!< Load texture compatible with the rules
                                        of KHR_texture_basisu glTF extension */
};
```
And this seems to be the one that is intended to be used there (because the documentation explicitly mentions `KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT`)

But still, there are  [several places in the `texturetests.cc`](https://github.com/KhronosGroup/KTX-Software/blob/7a45c4e6159c0815e4663a529a1e718136a16e39/tests/texturetests/texturetests.cc#L2581) where this function is called as 
```
        result = ktxTexture2_CreateFromMemory(ktxMemFile, ktxMemFileLen,
                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
                                              &texture);
```

The `KTX_TEXTURE_CREATE_ALLOC_STORAGE` is not part of the `ktxTextureCreateFlagBits`. It is part of the [`ktxTextureCreateStorageEnum`](https://github.com/KhronosGroup/KTX-Software/blob/7a45c4e6159c0815e4663a529a1e718136a16e39/include/ktx.h#L763).

I assume that the fact that this works is... only a coincidence, because `KTX_TEXTURE_CREATE_ALLOC_STORAGE` happens to have the same value as `KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT`. (And if that's the case: Hooray for C++'s helpless attempt to disguise arbitrary _values_ as "types"...). 

But this is either a bug in the tests, or a case of a function being distressingly under-documented.


---

Some context: [A build](https://github.com/KhronosGroup/KTX-Software/runs/23220059688) for the [JNI improvements](https://github.com/KhronosGroup/KTX-Software/pull/886) PR is failing in the [test for `createFromMemory`](https://github.com/KhronosGroup/KTX-Software/blob/2306747c4f505ef40c3b564afc6f23b3587c18be/interface/java_binding/src/test/java/org/khronos/ktx/test/KtxTexture2Test.java#L87), saying 
```
org.khronos.ktx.test.KtxTexture2Test
[ERROR] testCreateFromMemoryBasic  Time elapsed: 0.023 s  <<< ERROR!
org.khronos.ktx.KtxException: Operation not allowed in the current state.
	at org.khronos.ktx.test.KtxTexture2Test.testCreateFromMemoryBasic(KtxTexture2Test.java:87)
```
So I wanted to look up what I _might_ have done wrong there. This failure happens **only** on Linux (and **not** on Windows), so maybe it's not even my fault, but I'll dig into it...

--- a/lib/texture2.c
+++ b/lib/texture2.c
@@ -2494,6 +2494,7 @@ ktxTexture2_inflateZstdInt(ktxTexture2* This, ktx_uint8_t* pDeflatedData,
     ktxLevelIndexEntry* cindex = This->_private->_levelIndex;
     ktxLevelIndexEntry* nindex;
     ktx_uint32_t uncompressedLevelAlignment;
+    ktx_error_code_e result;
 
     ZSTD_DCtx* dctx;
 
@@ -2515,6 +2516,10 @@ ktxTexture2_inflateZstdInt(ktxTexture2* This, ktx_uint8_t* pDeflatedData,
 
     ktx_size_t inflatedByteLength = 0;
     dctx = ZSTD_createDCtx();
+    if (dctx == NULL) {
+        result = KTX_OUT_OF_MEMORY;
+        goto cleanup;
+    }
     for (int32_t level = This->numLevels - 1; level >= 0; level--) {
         size_t levelByteLength =
             ZSTD_decompressDCtx(dctx, pInflatedData + levelOffset,
@@ -2525,13 +2530,17 @@ ktxTexture2_inflateZstdInt(ktxTexture2* This, ktx_uint8_t* pDeflatedData,
             ZSTD_ErrorCode error = ZSTD_getErrorCode(levelByteLength);
             switch(error) {
               case ZSTD_error_dstSize_tooSmall:
-                return KTX_DECOMPRESS_LENGTH_ERROR; // inflatedDataCapacity too small.
+                result = KTX_DECOMPRESS_LENGTH_ERROR; // inflatedDataCapacity too small.
+                goto cleanup;
               case ZSTD_error_checksum_wrong:
-                return KTX_DECOMPRESS_CHECKSUM_ERROR;
+                result = KTX_DECOMPRESS_CHECKSUM_ERROR;
+                goto cleanup;
               case ZSTD_error_memory_allocation:
-                return KTX_OUT_OF_MEMORY;
-              default:
-                return KTX_FILE_DATA_ERROR;
+                result =  KTX_OUT_OF_MEMORY;
+                goto cleanup;
+             default:
+                result = KTX_FILE_DATA_ERROR;
+                goto cleanup;
             }
         }
 
@@ -2562,6 +2571,11 @@ ktxTexture2_inflateZstdInt(ktxTexture2* This, ktx_uint8_t* pDeflatedData,
     bdb[KHR_DF_WORD_BYTESPLANE0] = prtctd->_formatSize.blockSizeInBits / 8;
 
     return KTX_SUCCESS;
+
+cleanup:
+    ZSTD_freeDCtx(dctx);
+    free(nindex);
+    return result;
 }
 
 /**
@@ -2616,11 +2630,15 @@ ktxTexture2_inflateZLIBInt(ktxTexture2* This, ktx_uint8_t* pDeflatedData,
                                                     &levelByteLength,
                                                     &pDeflatedData[cindex[level].byteOffset],
                                                     cindex[level].byteLength);
-        if (result != KTX_SUCCESS)
+        if (result != KTX_SUCCESS) {
+            free(nindex);
             return result;
+        }
 
-        if (This->_private->_levelIndex[level].uncompressedByteLength != levelByteLength)
+        if (This->_private->_levelIndex[level].uncompressedByteLength != levelByteLength) {
+            free(nindex);
             return KTX_DECOMPRESS_LENGTH_ERROR;
+        }
 
         nindex[level].byteOffset = levelOffset;
         nindex[level].uncompressedByteLength = nindex[level].byteLength =

--- a/lib/writer2.c
+++ b/lib/writer2.c
@@ -782,8 +782,11 @@ ktxTexture2_DeflateZstd(ktxTexture2* This, ktx_uint32_t compressionLevel)
     ktxLevelIndexEntry* cindex = This->_private->_levelIndex;
     ktxLevelIndexEntry* nindex;
     ktx_uint8_t* pCmpDst;
+    ktx_error_code_e result;
 
     ZSTD_CCtx* cctx = ZSTD_createCCtx();
+    if (cctx == NULL)
+        return KTX_OUT_OF_MEMORY;
 
     if (This->supercompressionScheme != KTX_SS_NONE)
         return KTX_INVALID_OPERATION;
@@ -797,8 +800,10 @@ ktxTexture2_DeflateZstd(ktxTexture2* This, ktx_uint32_t compressionLevel)
     }
 
     workBuf = malloc(dstRemainingByteLength + levelIndexByteLength);
-    if (workBuf == NULL)
-        return KTX_OUT_OF_MEMORY;
+    if (workBuf == NULL) {
+        result = KTX_OUT_OF_MEMORY;
+        goto cleanup;
+    }
     nindex = (ktxLevelIndexEntry*)workBuf;
     pCmpDst = &workBuf[levelIndexByteLength];
 
@@ -814,28 +819,33 @@ ktxTexture2_DeflateZstd(ktxTexture2* This, ktx_uint32_t compressionLevel)
             ZSTD_ErrorCode error = ZSTD_getErrorCode(levelByteLengthCmp);
             switch(error) {
               case ZSTD_error_parameter_outOfBound:
-                return KTX_INVALID_VALUE;
+                result = KTX_INVALID_VALUE;
+                goto cleanup;
               case ZSTD_error_dstSize_tooSmall:
 #ifdef DEBUG
                 assert(false && "Deflate dstSize too small.");
 #else
-                return KTX_OUT_OF_MEMORY;
+                result = KTX_OUT_OF_MEMORY;
+                goto cleanup;
 #endif
               case ZSTD_error_workSpace_tooSmall:
 #ifdef DEBUG
                 assert(false && "Deflate workspace too small.");
 #else
-                return KTX_OUT_OF_MEMORY;
+                result = KTX_OUT_OF_MEMORY;
+                goto cleanup;
 #endif
               case ZSTD_error_memory_allocation:
-                return KTX_OUT_OF_MEMORY;
+                result = KTX_OUT_OF_MEMORY;
+                goto cleanup;
               default:
                 // The remaining errors look like they should only
                 // occur during decompression but just in case.
 #ifdef DEBUG
                 assert(true);
 #else
-                return KTX_INVALID_OPERATION;
+                result = KTX_INVALID_OPERATION;
+                goto cleanup;
 #endif
             }
         }
@@ -868,6 +878,11 @@ ktxTexture2_DeflateZstd(ktxTexture2* This, ktx_uint32_t compressionLevel)
     bdb[KHR_DF_WORD_BYTESPLANE0] = 0; /* bytesPlane3..0 = 0 */
 
     return KTX_SUCCESS;
+
+cleanup:
+    ZSTD_freeCCtx(cctx);
+    free(workBuf);
+    return result;
 }
 
 /**
@@ -919,8 +934,10 @@ ktxTexture2_DeflateZLIB(ktxTexture2* This, ktx_uint32_t compressionLevel)
                                                     &This->pData[cindex[level].byteOffset],
                                                     cindex[level].byteLength,
                                                     compressionLevel);
-        if (result != KTX_SUCCESS)
+        if (result != KTX_SUCCESS) {
+            free(workBuf);
             return result;
+        }
 
         nindex[level].byteOffset = levelOffset;
         nindex[level].uncompressedByteLength = cindex[level].byteLength;

--- a/tests/texturetests/texturetests.cc
+++ b/tests/texturetests/texturetests.cc
@@ -2579,12 +2579,12 @@ TEST_F(ktxTexture2_MetadataTest, EmptyValue) {
 
     if (ktxMemFile != NULL) {
         result = ktxTexture2_CreateFromMemory(ktxMemFile, ktxMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         result = ktxHashList_AddKVPair(&texture->kvDataHead,
                                        "MSCtestKey", 0, nullptr);
@@ -2599,12 +2599,12 @@ TEST_F(ktxTexture2_MetadataTest, EmptyValue) {
             ktxTexture_Destroy(ktxTexture(texture));
 
         result = ktxTexture2_CreateFromMemory(newMemFile, newMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         ktx_uint32_t valueLen;
         ktx_uint8_t* value;
@@ -2628,12 +2628,12 @@ TEST_F(ktxTexture2_MetadataTest, NoMetadata) {
 
     if (ktxMemFile != NULL) {
         result = ktxTexture2_CreateFromMemory(ktxMemFile, ktxMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         ktxHashList_Destruct(&texture->kvDataHead);
         ktxTexture(texture)->kvDataHead = nullptr;
@@ -2651,12 +2651,12 @@ TEST_F(ktxTexture2_MetadataTest, NoMetadata) {
             ktxTexture_Destroy(ktxTexture(texture));
 
         result = ktxTexture2_CreateFromMemory(newMemFile, newMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         ktx_uint32_t valueLen;
         ktx_uint8_t* value;
@@ -2678,12 +2678,12 @@ TEST_F(ktxTexture2_MetadataTest, NoLibVersionDupOnMultipleWrites) {
 
     if (ktxMemFile != NULL) {
         result = ktxTexture2_CreateFromMemory(ktxMemFile, ktxMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         const ktx_uint32_t iterations = 2;
         ktx_size_t newMemFileLens[iterations];
@@ -2707,12 +2707,12 @@ TEST_F(ktxTexture2_MetadataTest, NoLibVersionDupOnMultipleWrites) {
             ktx_uint8_t* value;
             result = ktxTexture2_CreateFromMemory(newMemFiles[i],
                                                   newMemFileLens[i],
-                                                  KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                                  KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                                   &texture);
             ASSERT_TRUE(result == KTX_SUCCESS);
             ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                          << ktxErrorString(result);
-            ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+            ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
             result = ktxHashList_FindValue(&texture->kvDataHead,
                                           "KTXwriter",
@@ -2746,12 +2746,12 @@ TEST_F(ktxTexture2_MetadataTest, LibVersionUpdatedCorrectly) {
 
     if (ktxMemFile != NULL) {
         result = ktxTexture2_CreateFromMemory(ktxMemFile, ktxMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         ktx_uint32_t curWriterLen;
         ktx_uint8_t* curWriterVal;
@@ -2792,12 +2792,12 @@ TEST_F(ktxTexture2_MetadataTest, LibVersionUpdatedCorrectly) {
         ktx_uint8_t* newWriterVal;
         result = ktxTexture2_CreateFromMemory(newMemFile,
                                               newMemFileLen,
-                                              KTX_TEXTURE_CREATE_ALLOC_STORAGE,
+                                              KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
                                               &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
         ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image data not loaded";
 
         result = ktxHashList_FindValue(&texture->kvDataHead,
                                        "KTXwriter",

--- a/tests/unittests/unittests.cc
+++ b/tests/unittests/unittests.cc
@@ -936,9 +936,9 @@ class SwizzleTestBase : public ::testing::Test {
                                     KTX_TEXTURE_CREATE_ALLOC_STORAGE,
                                     &texture);
         ASSERT_TRUE(result == KTX_SUCCESS);
-        ASSERT_TRUE(texture != NULL) << "ktxTexture_CreateFromMemory failed: "
+        ASSERT_TRUE(texture != NULL) << "ktxTexture_Create failed: "
                                      << ktxErrorString(result);
-        ASSERT_TRUE(texture->pData != NULL) << "Image stoage not allocated";
+        ASSERT_TRUE(texture->pData != NULL) << "Image storage not allocated";
 
         result = helper.copyImagesToTexture(ktxTexture(texture));
         ASSERT_TRUE(result == KTX_SUCCESS);

